### PR TITLE
Add ERC: Code Index

### PR DESCRIPTION
**Pull request by peersky**

### CODE REVIEW

File: ERCS/erc-0000.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+eip: erc-0000.md
```
Comment (abcoathup) [inline ERCS/erc-0000.md:2]:
```suggestion
eip: 7744
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

File: ERCS/erc-0000.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+title: Code Index
+description: Indexing smart contract bytecode on Ethereum.
+author: Tim Pechersky (@peersky) <t@peersky.xyz>
+discussions-to: https://ethereum-magicians.org/t/erc-code-index/20569
```
Comment (abcoathup) [inline ERCS/erc-0000.md:6]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7744-code-index/20569
```

**Comment by peersky** [comment]
@abcoathup Im not really sure how to fix the [HTMLProofer CI error](https://github.com/ethereum/ERCs/actions/runs/9993266603/job/27620237667?pr=542), it seems that it might look for a wrong directory path, could you let me know if any action needed from me?

File: ERCS/erc-7744.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+**Bytecode over Addresses**: Bytecode is deterministic and can be verified on-chain, while addresses are opaque and mutable.
+
+**Reverting on re-indexing**: There is small, yet non-zero probability of hash collision attack. Disallowing updates to indexed location of bytecode coupes with this.
+
+**Simple Interface**: The interface is minimal and focused to maximize composability and ease of implementation.
+
+## Reference Implementation
+
+- [../assets/erc-7744/](../assets/erc-7744/)
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:91]:
```suggestion
- [Reference Implementation](../assets/eip-7744/README.md)
```

File: ERCS/erc-7744.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
+---
+eip: 7744
+title: Code Index
+description: Indexing smart contract bytecode on Ethereum.
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:4]:
```suggestion
title: Code Index
description: Indexing smart contract bytecode
```

In addition to removing the fluff, you should expand a little more on what this proposal proposes. What is a code index? Who would want it? Try not to just restate your title in your description. Instead, elaborate in the description on the ideas introduced in the title.

File: ERCS/erc-7744.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+## Abstract
+
+This EIP defines a standard interface for indexing smart contracts on Ethereum by their bytecode hash. This enables trustless discovery and verification of contract code, facilitating use cases like bytecode signing, whitelisting, and decentralized distribution mechanisms.
+
+## Motivation
+
+Existing contract discovery relies on addresses, which are non-deterministic and can be obfuscated through proxies. Indexing by bytecode hash provides a deterministic and tamper-proof way to identify and verify contract code, enhancing security and trust in the Ethereum ecosystem.
+
+Global bytecode identification is a requirement to build efficient at high scale global factories for future software distribution, that may want to rely on extensive re-use of same codebase as own dependencies.
+
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:22]: I'd like to see a bit more of an in-depth usecase here. Walk the reader through a concrete example of why someone would want to record code hashes on chain.

File: ERCS/erc-7744.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     * @dev returns zero if the contract is not indexed
+     * @param id The bytecode hash
+     * @return The contract address
+     */
+    function get(bytes32 id) external view returns (address) {
+        return index[id];
+    }
+}
+
+```
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:75]: We generally don't allow implementations in the specification section. We have made exceptions for registries in the past, but you'll need to explain (in the document) why this registry needs to be an EIP at all, and not just a library you publish.

File: ERCS/erc-7744.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+The `CodeIndex` contract is deployed at: `0xc0D31d398c5ee86C5f8a23FA253ee8a586dA03Ce` using `CREATE2` via the deterministic deployer at `0x4e59b44847b379578588920ca78fbf26c0b4956c` with a salt of `0x220a70730c743a005cfd55180805d2c0d5b8c7695c5496100dcffa91c02befce`.
+
+## Rationale
+
+**Bytecode over Addresses**: Bytecode is deterministic and can be verified on-chain, while addresses are opaque and mutable.
+
+**Reverting on re-indexing**: There is small, yet non-zero probability of hash collision attack. Disallowing updates to indexed location of bytecode coupes with this.
+
+**Simple Interface**: The interface is minimal and focused to maximize composability and ease of implementation.
+
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:88]: You should explain where you got your salt from. See https://en.wikipedia.org/wiki/Nothing-up-my-sleeve_number.

File: assets/erc-7744/abi/src/CodeIndex.sol/CodeIndex.json
Comment (SamWilsn) [inline assets/erc-7744/abi/src/CodeIndex.sol/CodeIndex.json:1]: Please remove files that aren't directly related to the reference implementation. This looks like it was generated from the solidity?

File: assets/erc-7744/abi/src/ICodeIndex.sol/ICodeIndex.json
Comment (SamWilsn) [inline assets/erc-7744/abi/src/ICodeIndex.sol/ICodeIndex.json:1]: Same here.

File: assets/erc-7744/abi/src/TestFacet.sol/TestFacet.json
Comment (SamWilsn) [inline assets/erc-7744/abi/src/TestFacet.sol/TestFacet.json:1]: And again.

File: assets/erc-7744/hardhat.config.ts
Comment (SamWilsn) [inline assets/erc-7744/hardhat.config.ts:1]: Please remove build system files.

File: assets/erc-7744/package.json
Comment (SamWilsn) [inline assets/erc-7744/package.json:1]: More build system files.

File: assets/erc-7744/src/TestFacet.sol
Comment (SamWilsn) [inline assets/erc-7744/src/TestFacet.sol:1]: Not sure how relevant this is?

File: assets/erc-7744/yarn.lock
Comment (SamWilsn) [inline assets/erc-7744/yarn.lock:1]: Definitely remove this file.

**Review by SamWilsn — CHANGES_REQUESTED** [review]

File: ERCS/erc-7744.md (hunk 6 above)
Comment (peersky) [reply ERCS/erc-7744.md:90]: added motivation

File: ERCS/erc-7744.md (hunk 7 above)
Comment (peersky) [reply ERCS/erc-7744.md:103]: added explanation

File: ERCS/erc-7744.md (hunk 5 above)
Comment (peersky) [reply ERCS/erc-7744.md:37]: added motivation

File: ERCS/erc-7744.md (hunk 4 above)
Comment (peersky) [reply ERCS/erc-7744.md]: Elaborated description. Not sure if there is any length requirement on this field. Im concerned if it's too long now?

File: ERCS/erc-7744.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+## Motivation
+
+Existing contract discovery relies on addresses, which are non-deterministic and can be obfuscated through proxies. Indexing by bytecode hash provides a deterministic and tamper-proof way to identify and verify contract code, enhancing security and trust in the Ethereum ecosystem.
+
+Consider a security auditor who wants to attest to the integrity of a contract's code. By referencing bytecode hashes, auditors can focus their audit on the bytecode itself, without needing to assess deployment parameters or storage contents. This method verifies the integrity of a contract's codebase without auditing the entire contract state.
+
+Additionally, a bytecode referencing allows whitelist contracts before deployment, allowing developers to get pre-approval for their codebase without disclosing the code itself, or even pre-setup infrastructure that will change it behavior upon adding some determined functionality on chain.
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:23]:
```suggestion
Additionally, bytecode referencing allows whitelist contracts before deployment, allowing developers to get pre-approval for their codebase without disclosing the code itself, or even pre-setup infrastructure that will change it behavior upon adding some determined functionality on chain.
```

File: ERCS/erc-7744.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+**Simple Interface**: The interface is minimal and focused to maximize composability and ease of implementation.
+
+**Library Implementation**: Implementing this as a library would limit its impact, making code reuse more difficult and lacking a single, official source of truth. By establishing this as an ERC, we ensure standardization and widespread adoption, driving the ecosystem forward.
+
+## Reference Implementation
+
+Reference implementation of the Code Index can be found in the assets folder. There you can find the [interface](../assets/eip-7744/ICodeIndex.sol) and the [implementation](../assets/eip-7744/CodeIndex.sol) of the Code Index.
+
+For mor thorough implementation, you can refer to the Ethereum Distribution System (EDS):
+`https://github.com/peeramid-labs/eds`
```
Comment (SamWilsn) [inline ERCS/erc-7744.md:111]:
```suggestion
```

Please remove this external link.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**